driver/src/main/com/mongodb/MongoCollectionImpl.java

```diff
@@ -217,8 +217,38 @@ public long count(final Bson filter, final CountOptions options) {
 
     @Override
     public <TResult> FindIterable<TResult> find(final Bson filter, final Class<TResult> resultClass) {
-        return new FindIterableImpl<TDocument, TResult>(namespace, this.documentClass, resultClass, codecRegistry, readPreference,
-                readConcern, executor, filter, new FindOptions());
+        return createFindIterable(null, filter, resultClass);
+    }
+
+    @Override
+    public FindIterable<TDocument> find(final ClientSession clientSession) {
+        notNull("clientSession", clientSession);
```

Comment:
Checking notNull here seems odd, as its chained you don't need it, it gets checked later.

I'm not 100% sold on allowing `ClientSession` to be null at all. Could we mimic the `NoOpSessionContext` and ensure we always have a `ClientSession`?  If we did that we can chain the methods as we have traditionally done and theres no need for the private createIterable / execute methods.

Reply:
The notNull check is a bit of defensive programming in case the class is refactored later such that this method no longer calls another public method that takes care of the null check.  But I can add a unit test for that instead and remove the redundant check.

I don't think we can mimic NoOpSessionContext because at some level we have to know that the application did not supply their own session, and therefore create the implicit session (this currently happens in OperationExecutor#getClientSession).  Checking for null is a convenient way to do that.

I'm not clear on the relationship between allowing a null ClientSession and method chaining for iterables.  Can you elaborate?